File: generate_catalog.py
#!/usr/bin/env python3
"""
Build the library catalog from your EPUB files.

HOW TO USE
----------
1. Put your .epub files inside the "epubs" folder.
   - To group them into collections (Prayer, Leadership, Marriage, ...),
     make sub-folders and drop the books inside, e.g.:
         epubs/Prayer/how-to-pray.epub
         epubs/Leadership/the-art-of-leadership.epub
     The sub-folder name becomes the collection shown in the app.
   - Books placed directly in "epubs" (no sub-folder) go under "All books".
2. Run:  python3 generate_catalog.py
3. It writes books.json and fills the covers/ folder. That's it.

Only the Python standard library is used - nothing to install.
"""

import json
import os
import re
import sys
import zipfile
import xml.etree.ElementTree as ET

ROOT = os.path.dirname(os.path.abspath(__file__))
EPUB_DIR = os.path.join(ROOT, "epubs")
COVER_DIR = os.path.join(ROOT, "covers")
OUT = os.path.join(ROOT, "books.json")

IMG_EXT = {"image/jpeg": ".jpg", "image/jpg": ".jpg", "image/png": ".png",
           "image/gif": ".gif", "image/webp": ".webp", "image/svg+xml": ".svg"}


def local(tag):
    return tag.rsplit("}", 1)[-1].lower()


def slugify(name):
    s = re.sub(r"[^a-zA-Z0-9]+", "-", name).strip("-").lower()
    return s or "book"


def find_opf_path(z):
    with z.open("META-INF/container.xml") as f:
        root = ET.parse(f).getroot()
    for el in root.iter():
        if local(el.tag) == "rootfile" and el.get("full-path"):
            return el.get("full-path")
    return None


def parse_epub(path):
    """Return (title, author, cover_bytes, cover_ext) - any may be None."""
    with zipfile.ZipFile(path) as z:
        opf_path = find_opf_path(z)
        if not opf_path:
            return None, None, None, None
        opf_dir = os.path.dirname(opf_path)
        with z.open(opf_path) as f:
            opf = ET.parse(f).getroot()

        title = author = None
        cover_id = None
        manifest = {}          # id -> (href, media_type, properties)
        href_by_props = None

        for el in opf.iter():
            t = local(el.tag)
            if t == "title" and title is None and (el.text or "").strip():
                title = el.text.strip()
            elif t == "creator" and author is None and (el.text or "").strip():
                author = el.text.strip()
            elif t == "meta" and el.get("name") == "cover":
                cover_id = el.get("content")
            elif t == "item":
                iid = el.get("id"); href = el.get("href")
                mt = (el.get("media-type") or "").lower()
                props = (el.get("properties") or "").lower()
                if iid and href:
                    manifest[iid] = (href, mt, props)
                    if "cover-image" in props:
                        href_by_props = (href, mt)

        # resolve cover href, in priority order
        cover_href = cover_mt = None
        if href_by_props:
            cover_href, cover_mt = href_by_props
        elif cover_id and cover_id in manifest:
            cover_href, cover_mt, _ = manifest[cover_id]
        else:
            for href, mt, props in manifest.values():
                if mt.startswith("image/") and "cover" in href.lower():
                    cover_href, cover_mt = href, mt
                    break

        cover_bytes = cover_ext = None
        if cover_href:
            full = os.path.normpath(os.path.join(opf_dir, cover_href)).replace("\\", "/")
            try:
                cover_bytes = z.read(full)
                cover_ext = IMG_EXT.get(cover_mt) or os.path.splitext(full)[1] or ".jpg"
            except KeyError:
                cover_bytes = None
        return title, author, cover_bytes, cover_ext


def main():
    if not os.path.isdir(EPUB_DIR):
        print("No 'epubs' folder found. Create it and add your .epub files.")
        sys.exit(1)
    os.makedirs(COVER_DIR, exist_ok=True)

    books = []
    seen = set()
    for dirpath, _, files in os.walk(EPUB_DIR):
        for fn in sorted(files):
            if not fn.lower().endswith(".epub"):
                continue
            full = os.path.join(dirpath, fn)
            rel = os.path.relpath(full, ROOT).replace("\\", "/")
            # collection = immediate sub-folder under epubs/, else "All books"
            inside = os.path.relpath(dirpath, EPUB_DIR).replace("\\", "/")
            collection = "All books" if inside in (".", "") else inside.split("/")[0]

            base = os.path.splitext(fn)[0]
            bid = slugify(base)
            n = bid
            i = 2
            while n in seen:
                n = f"{bid}-{i}"; i += 1
            bid = n; seen.add(bid)

            try:
                title, author, cover_bytes, cover_ext = parse_epub(full)
            except Exception as e:
                print(f"  ! Skipped {fn}: {e}")
                continue

            cover_rel = None
            if cover_bytes:
                cover_name = bid + (cover_ext if cover_ext.startswith(".") else ".jpg")
                with open(os.path.join(COVER_DIR, cover_name), "wb") as cf:
                    cf.write(cover_bytes)
                cover_rel = "covers/" + cover_name

            books.append({
                "id": bid,
                "title": title or base.replace("-", " ").title(),
                "author": author or "",
                "collection": collection,
                "file": rel,
                "cover": cover_rel,
            })
            print(f"  \u2713 {books[-1]['title']}  ({collection})")

    books.sort(key=lambda b: (b["collection"].lower(), b["title"].lower()))
    with open(OUT, "w", encoding="utf-8") as f:
        json.dump({"books": books}, f, ensure_ascii=False, indent=2)
    print(f"\nDone. {len(books)} book(s) written to books.json.")
    if not books:
        print("Tip: put .epub files inside the 'epubs' folder, then run this again.")


if __name__ == "__main__":
    main()

File: generate_catalog.R
#!/usr/bin/env Rscript
# ============================================================
# The Reading Room - build the library catalog, in R
# ------------------------------------------------------------
# HOW TO USE
#   1. Put your .epub files in the "epubs" folder
#      (optionally in sub-folders named by collection,
#       e.g. epubs/Prayer/, epubs/Leadership/).
#   2. Open R/RStudio IN this app folder, or set it as the
#      working directory, then run:
#            source("generate_catalog.R")
#      (or from a terminal in this folder:  Rscript generate_catalog.R )
#   3. It writes books.json and fills the covers/ folder.
#
# Uses ONLY base R - nothing to install.
# ============================================================

root      <- getwd()
epub_dir  <- file.path(root, "epubs")
cover_dir <- file.path(root, "covers")
out_file  <- file.path(root, "books.json")

if (!dir.exists(epub_dir)) {
  stop("No 'epubs' folder found in: ", root,
       "\nOpen R inside the app folder (or setwd() to it) and try again.")
}
if (!dir.exists(cover_dir)) dir.create(cover_dir)

img_ext <- c("image/jpeg"=".jpg","image/jpg"=".jpg","image/png"=".png",
             "image/gif"=".gif","image/webp"=".webp","image/svg+xml"=".svg")

## ---- small helpers ----
read_zip_raw <- function(zip, name) {
  con <- unz(zip, name, open = "rb"); on.exit(close(con))
  out <- raw(0)
  repeat {
    chunk <- readBin(con, "raw", n = 1048576L)
    if (length(chunk) == 0L) break
    out <- c(out, chunk)
  }
  out
}
read_zip_text <- function(zip, name) {
  txt <- rawToChar(read_zip_raw(zip, name)); Encoding(txt) <- "UTF-8"; txt
}
get_attr <- function(tag, name) {
  m <- regexec(paste0(name, "\\s*=\\s*\"([^\"]*)\""), tag, perl = TRUE)
  g <- regmatches(tag, m)[[1]]
  if (length(g) >= 2) g[2] else NA_character_
}
first_tag_text <- function(xml, tag) {
  m <- regexec(paste0("(?s)<", tag, "[^>]*>\\s*(.*?)\\s*</", tag, ">"), xml, perl = TRUE)
  g <- regmatches(xml, m)[[1]]
  if (length(g) >= 2 && nzchar(g[2])) g[2] else NA_character_
}
unescape <- function(s) {
  if (is.na(s)) return(s)
  s <- gsub("&amp;",  "&", s, fixed = TRUE); s <- gsub("&lt;", "<", s, fixed = TRUE)
  s <- gsub("&gt;",   ">", s, fixed = TRUE); s <- gsub("&quot;","\"", s, fixed = TRUE)
  s <- gsub("&#39;",  "'", s, fixed = TRUE); s <- gsub("&apos;","'", s, fixed = TRUE)
  s <- gsub("&#160;", " ", s, fixed = TRUE); s <- gsub("&nbsp;"," ", s, fixed = TRUE)
  s
}
slugify <- function(x) {
  x <- tolower(x); x <- gsub("[^a-z0-9]+", "-", x); x <- gsub("^-+|-+$", "", x)
  if (nchar(x) == 0) "book" else x
}
norm_path <- function(p) {
  parts <- strsplit(p, "/", fixed = TRUE)[[1]]; out <- character(0)
  for (seg in parts) {
    if (seg == "" || seg == ".") next
    else if (seg == "..") { if (length(out)) out <- out[-length(out)] }
    else out <- c(out, seg)
  }
  paste(out, collapse = "/")
}
json_str <- function(s) {
  if (is.na(s)) return("\"\"")
  s <- gsub("\\", "\\\\", s, fixed = TRUE); s <- gsub("\"", "\\\"", s, fixed = TRUE)
  s <- gsub("\r", "", s, fixed = TRUE);     s <- gsub("\n", "\\n", s, fixed = TRUE)
  s <- gsub("\t", "\\t", s, fixed = TRUE)
  paste0("\"", s, "\"")
}

## ---- parse one epub ----
parse_epub <- function(path) {
  container <- read_zip_text(path, "META-INF/container.xml")
  opf_path  <- get_attr(container, "full-path")
  if (is.na(opf_path)) return(NULL)
  opf     <- read_zip_text(path, opf_path)
  opf_dir <- dirname(opf_path); if (opf_dir == ".") opf_dir <- ""

  title  <- unescape(first_tag_text(opf, "dc:title"))
  author <- unescape(first_tag_text(opf, "dc:creator"))

  items <- regmatches(opf, gregexpr("<item\\b[^>]*>", opf, perl = TRUE))[[1]]
  href_by_id <- list(); cover_href <- NA_character_; cover_mt <- NA_character_
  for (it in items) {
    id    <- get_attr(it, "id");  href <- get_attr(it, "href")
    mt    <- get_attr(it, "media-type"); props <- get_attr(it, "properties")
    if (!is.na(id) && !is.na(href)) href_by_id[[id]] <- list(href = href, mt = mt)
    if (!is.na(props) && grepl("cover-image", props) && is.na(cover_href)) {
      cover_href <- href; cover_mt <- mt
    }
  }
  if (is.na(cover_href)) {                       # <meta name="cover" content="ID">
    metas <- regmatches(opf, gregexpr("<meta\\b[^>]*>", opf, perl = TRUE))[[1]]
    for (mt in metas) if (identical(get_attr(mt, "name"), "cover")) {
      cid <- get_attr(mt, "content")
      if (!is.na(cid) && !is.null(href_by_id[[cid]])) {
        cover_href <- href_by_id[[cid]]$href; cover_mt <- href_by_id[[cid]]$mt
      }
    }
  }
  if (is.na(cover_href)) {                        # last resort: an image named "cover"
    for (v in href_by_id) if (!is.na(v$mt) && startsWith(v$mt, "image/") &&
                              grepl("cover", v$href, ignore.case = TRUE)) {
      cover_href <- v$href; cover_mt <- v$mt; break
    }
  }

  cover_bytes <- NULL; cover_ext <- NA_character_
  if (!is.na(cover_href)) {
    full <- if (nzchar(opf_dir)) norm_path(paste0(opf_dir, "/", cover_href)) else norm_path(cover_href)
    cover_bytes <- tryCatch(read_zip_raw(path, full), error = function(e) NULL)
    if (!is.null(cover_bytes)) {
      ext <- tolower(tools::file_ext(cover_href))
      cover_ext <- if (nzchar(ext)) paste0(".", ext)
                   else if (!is.na(cover_mt) && cover_mt %in% names(img_ext)) img_ext[[cover_mt]]
                   else ".jpg"
    }
  }
  list(title = title, author = author, cover_bytes = cover_bytes, cover_ext = cover_ext)
}

## ---- walk the epubs folder ----
files <- list.files(epub_dir, pattern = "\\.epub$", recursive = TRUE,
                    full.names = TRUE, ignore.case = TRUE)
files <- sort(files)
records <- list(); seen <- character(0)

for (f in files) {
  rel_root  <- substring(f, nchar(root) + 2L)            # e.g. epubs/Prayer/x.epub
  rel_epub  <- substring(f, nchar(epub_dir) + 2L)        # e.g. Prayer/x.epub
  parts     <- strsplit(rel_epub, "/", fixed = TRUE)[[1]]
  collection <- if (length(parts) >= 2) parts[1] else "All books"

  base <- tools::file_path_sans_ext(basename(f))
  id <- slugify(base); n <- id; i <- 2
  while (n %in% seen) { n <- paste0(id, "-", i); i <- i + 1 }
  id <- n; seen <- c(seen, id)

  info <- tryCatch(parse_epub(f), error = function(e) { message("  ! Skipped ", basename(f), ": ", conditionMessage(e)); NULL })
  if (is.null(info)) next

  cover_rel <- NA_character_
  if (!is.null(info$cover_bytes)) {
    cover_name <- paste0(id, if (startsWith(info$cover_ext, ".")) info$cover_ext else ".jpg")
    writeBin(info$cover_bytes, file.path(cover_dir, cover_name))
    cover_rel <- paste0("covers/", cover_name)
  }
  title <- if (is.na(info$title)) tools::toTitleCase(gsub("-", " ", base)) else info$title

  records[[length(records) + 1L]] <- list(
    id = id, title = title, author = if (is.na(info$author)) "" else info$author,
    collection = collection, file = rel_root, cover = cover_rel)
  message("  \u2713 ", title, "  (", collection, ")")
}

## ---- sort + write books.json ----
if (length(records)) {
  ord <- order(tolower(vapply(records, `[[`, "", "collection")),
               tolower(vapply(records, `[[`, "", "title")))
  records <- records[ord]
}
obj_json <- function(b) paste0(
  "    {\n",
  "      \"id\": ",         json_str(b$id),         ",\n",
  "      \"title\": ",      json_str(b$title),      ",\n",
  "      \"author\": ",     json_str(b$author),     ",\n",
  "      \"collection\": ", json_str(b$collection), ",\n",
  "      \"file\": ",       json_str(b$file),       ",\n",
  "      \"cover\": ",       if (is.na(b$cover)) "null" else json_str(b$cover), "\n",
  "    }")
body <- paste(vapply(records, obj_json, ""), collapse = ",\n")
writeLines(paste0("{\n  \"books\": [\n", body, "\n  ]\n}"), out_file, useBytes = TRUE)

message("\nDone. ", length(records), " book(s) written to books.json.")
if (!length(records)) message("Tip: put .epub files in the 'epubs' folder, then run this again.")
